packages/go_router/lib/src/route_data.dart

```diff
@@ -205,6 +197,109 @@ abstract class ShellRouteData extends RouteData {
+abstract class StatefulShellRouteData extends RouteData {
+  /// Default const constructor
+  const StatefulShellRouteData();
+
+  /// [pageBuilder] is used to build the page
+  Page<void> pageBuilder(
+    BuildContext context,
+    GoRouterState state,
+    StatefulNavigationShell navigationShell,
+  ) =>
+      const NoOpPage();
+
+  /// [builder] is used to build the widget
+  Widget builder(
+    BuildContext context,
+    GoRouterState state,
+    StatefulNavigationShell navigationShell,
+  ) =>
+      throw UnimplementedError(
+        'One of `builder` or `pageBuilder` must be implemented.',
+      );
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellRoute $route<T extends StatefulShellRouteData>({
+    required T Function(GoRouterState) factory,
+    required List<StatefulShellBranch> branches,
+    ShellNavigationContainerBuilder? navigatorContainerBuilder,
+    String? restorationScopeId,
+  }) {
+    T factoryImpl(GoRouterState state) {
+      return (_stateObjectExpando[state] ??= factory(state)) as T;
+    }
+
+    Widget builder(
+      BuildContext context,
+      GoRouterState state,
+      StatefulNavigationShell navigationShell,
+    ) =>
+        factoryImpl(state).builder(
+          context,
+          state,
+          navigationShell,
+        );
+
+    Page<void> pageBuilder(
+      BuildContext context,
+      GoRouterState state,
+      StatefulNavigationShell navigationShell,
+    ) =>
+        factoryImpl(state).pageBuilder(
+          context,
+          state,
+          navigationShell,
+        );
+
+    if (navigatorContainerBuilder != null) {
+      return StatefulShellRoute(
+        branches: branches,
+        builder: builder,
+        pageBuilder: pageBuilder,
+        navigatorContainerBuilder: navigatorContainerBuilder,
+      );
+    }
+    return StatefulShellRoute.indexedStack(
+      branches: branches,
+      builder: builder,
+      pageBuilder: pageBuilder,
+      restorationScopeId: restorationScopeId,
+    );
+  }
+
+  /// Used to cache [StatefulShellRouteData] that corresponds to a given [GoRouterState]
+  /// to minimize the number of times it has to be deserialized.
+  static final Expando<StatefulShellRouteData> _stateObjectExpando =
+      Expando<StatefulShellRouteData>(
+    'GoRouteState to StatefulShellRouteData expando',
+  );
+}
+
+/// Base class for supporting
+/// [StatefulShellRoute](https://pub.dev/documentation/go_router/latest/go_router/StatefulShellRoute-class.html)
+abstract class StatefulShellBranchData {
+  /// Default const constructor
+  const StatefulShellBranchData();
+
+  /// A helper function used by generated code.
+  ///
+  /// Should not be used directly.
+  static StatefulShellBranch $branch<T extends StatefulShellBranchData>({
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) {
+    return StatefulShellBranch(
+      routes: routes,
```

Comment:
how does one specify initialLocation and restorationScopeId? we may also want a way to specify observers

Reply:
done